editor/plugins/gizmos/particles_3d_emission_shape_gizmo_plugin.cpp

```diff
@@ -0,0 +1,747 @@
+#include "scene/3d/gpu_particles_3d.h"
+#include "scene/resources/3d/primitive_meshes.h"
+#include "scene/resources/particle_process_material.h"
+
+Particles3DEmissionShapeGizmoPlugin::Particles3DEmissionShapeGizmoPlugin() {
+	helper.instantiate();
+
+	Color gizmo_color = EDITOR_DEF_RST("editors/3d_gizmos/gizmo_colors/particles_emission_shape", Color(0.5, 0.7, 1));
+	create_material("particles_emission_shape_material", gizmo_color);
+
+	create_handle_material("handles");
+}
+
+bool Particles3DEmissionShapeGizmoPlugin::has_gizmo(Node3D *p_spatial) {
+	return Object::cast_to<GPUParticles3D>(p_spatial) || Object::cast_to<CPUParticles3D>(p_spatial) != nullptr;
+}
+
+String Particles3DEmissionShapeGizmoPlugin::get_gizmo_name() const {
+	return "Particles3DEmissionShape";
+}
+
+int Particles3DEmissionShapeGizmoPlugin::get_priority() const {
+	return -1;
+}
+
+bool Particles3DEmissionShapeGizmoPlugin::is_selectable_when_hidden() const {
+	return true;
+}
+
+String Particles3DEmissionShapeGizmoPlugin::get_handle_name(const EditorNode3DGizmo *p_gizmo, int p_id, bool p_secondary) const {
+	if (Object::cast_to<GPUParticles3D>(p_gizmo->get_node_3d())) {
+		GPUParticles3D *particles = Object::cast_to<GPUParticles3D>(p_gizmo->get_node_3d());
+		Ref<ParticleProcessMaterial> mat = particles->get_process_material();
+		ParticleProcessMaterial::EmissionShape shape = mat->get_emission_shape();
+
+		if (shape == ParticleProcessMaterial::EMISSION_SHAPE_SPHERE || shape == ParticleProcessMaterial::EMISSION_SHAPE_SPHERE_SURFACE) {
+			return "Radius";
+		} else if (shape == ParticleProcessMaterial::EMISSION_SHAPE_BOX) {
+			switch (p_id) {
+				case 0:
+				case 1:
+					return "Size X";
+				case 2:
+				case 3:
+					return "Size Y";
+				case 4:
+				case 5:
+					return "Size Z";
+			}
+			return "";
+		} else if (shape == ParticleProcessMaterial::EMISSION_SHAPE_RING) {
+			if (p_id == 0) {
+				return "Ring Axis";
+			} else if (p_id == 1) {
+				return "Ring Height";
+			} else if (p_id == 2) {
+				return "Ring Radius";
+			} else {
+				return "Ring Inner Radius";
+			}
+		}
+	} else if (Object::cast_to<CPUParticles3D>(p_gizmo->get_node_3d())) {
+		CPUParticles3D *particles = Object::cast_to<CPUParticles3D>(p_gizmo->get_node_3d());
+		CPUParticles3D::EmissionShape shape = particles->get_emission_shape();
+
+		if (shape == CPUParticles3D::EMISSION_SHAPE_SPHERE || shape == CPUParticles3D::EMISSION_SHAPE_SPHERE_SURFACE) {
+			return "Radius";
+		} else if (shape == CPUParticles3D::EMISSION_SHAPE_BOX) {
+			switch (p_id) {
+				case 0:
+				case 1:
+					return "Size X";
+				case 2:
+				case 3:
+					return "Size Y";
+				case 4:
+				case 5:
+					return "Size Z";
+			}
+			return "";
+		} else if (shape == CPUParticles3D::EMISSION_SHAPE_RING) {
+			if (p_id == 0) {
+				return "Ring Axis";
+			} else if (p_id == 1) {
+				return "Ring Height";
+			} else if (p_id == 2) {
+				return "Ring Radius";
+			} else {
+				return "Ring Inner Radius";
```

Comment:
You don't prepend any of the other shape's handle names with ring, so it would probably be best if you could remove the "Ring " from these handle names as well.

Reply:
Fixed now, thanks for the feedback